Updated docstring in TimeSeries

davnish opened the pull request:
This solves the Issue #4168.
This updates the docstring for Timeseries. 
This will help us to see us the supported source packages for TimeSeries.

nabobalis commented:
Thank you for the PR @davnish, unfortunately this does not work. 

The names are actually to the `source` keyword and not the inputs line.

The issue with this approach is if we add more sources, we would have to manually change the text every time.

davnish commented:
@nabobalis how should I approach this issue then? Is there a way to create a dynamic docstring, I seached on the web but didnt find anything.

davnish commented:
@nabobalis Sorry for the delay, I have updated my approach. I have added an `__all__` in `__intit__.py` of `sunpy.timeseries.sources` which will keep track of the data source packages and when we will add a new source we can just update the `__all__`. will this work?

nabobalis commented on sunpy/timeseries/sources/__init__.py:
```diff
@@ -13,3 +13,7 @@
 from sunpy.timeseries.sources.norh import NoRHTimeSeries
 from sunpy.timeseries.sources.rhessi import RHESSISummaryTimeSeries
+
+__all__ = ['ESPTimeSeries', 'EVESpWxTimeSeries', 'GBMSummaryTimeSeries', 'XRSTimeSeries', 
```
I think adding this is fine.

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -46,9 +47,8 @@ class TimeSeriesFactory(BasicRegistrationFactory):
-    \\*inputs
-        Inputs to parse for timeseries objects. See the example section for a
-        detailed list of possible inputs.
+    inputs : {list_packages}
```
This is the wrong keyword. This needs to be reverted back to what it was.

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -46,9 +47,8 @@ class TimeSeriesFactory(BasicRegistrationFactory):
-        Inputs to parse for timeseries objects. See the example section for a
-        detailed list of possible inputs.
+    inputs : {list_packages}
+        This is a collection of data source packages currently supported by TimeSeries.
```
This is the wrong comment for this keyword

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -521,7 +521,8 @@ class NoTimeSeriesFound(ValueError):
                                default_widget_type=GenericTimeSeries,
                                additional_validation_functions=['is_datasource_for'])
+
+TimeSeriesFactory.__doc__ = TimeSeriesFactory.__doc__.format(list_packages = input_lines)
```
I think this is ok. I would prefer if we can follow the style we use for the time types formatting.

nabobalis commented:
> @nabobalis Sorry for the delay, I have updated my approach. I have added an `__all__` in `__intit__.py` of `sunpy.timeseries.sources` which will keep track of the data source packages and when we will add a new source we can just update the `__all__`. will this work?

It helps but the class names do not match the value used by the source keyword. We would need to extract them.

davnish commented:
Okay I was confused, I thought that the class names were the input lines. How should I extract the value used by source keywords?

nabobalis commented:
I'm not sure where the source is stored in each class, you'll have to check and see if we can access it

davnish commented:
Okay, I cross checked,
`['eve', 'esp' , 'gbmsummary', 'xrs', 'lyra', 'noaaindices', 'noaapredictindices', 'norh', 'rhessi']
`
These are sources currently supported. where should I add this list? and I will change the `inputs` keyword to `sources`

nabobalis commented:
They go with the source keyword. Ideally they should be added to the type location following the numpydoc style.

davnish commented:
> They go with the source keyword. Ideally they should be added to the type location following the numpydoc style.

Okay, I will add them to type location following the numpy docstyle. but wont this approach make the docstring static?

davnish commented:
> The issue with this approach is if we add more sources, we would have to manually change the text every time.

Okay, I will add them to type location following the numpy docstyle. but wont this approach make the docstring static?

nabobalis commented:
> > The issue with this approach is if we add more sources, we would have to manually change the text every time.
> 
> Okay, I will add them to type location following the numpy docstyle. but wont this approach make the docstring static?

Yes but you'll need to work out a method to make that dynamic.

davnish commented:
I have added a source list in `init.py`, which we can update when we add a new source. Are these changes right to do?

nabobalis commented on sunpy/timeseries/sources/__init__.py:
```diff
@@ -13,3 +13,5 @@
 from sunpy.timeseries.sources.norh import NoRHTimeSeries
 from sunpy.timeseries.sources.rhessi import RHESSISummaryTimeSeries
+
+source_list = ['eve', 'esp', 'gbmsummary', 'xrs', 'lyra', 'noaaindices', 'noaapredictindices', 'norh', 'rhessi']
```
Can we not create this dynamically?

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -46,11 +47,10 @@ class TimeSeriesFactory(BasicRegistrationFactory):
-    \\*inputs
-        Inputs to parse for timeseries objects. See the example section for a
-        detailed list of possible inputs.
+    inputs : esp, eve, gbmsummary, lyra, noaaindices, noaapredictindices, norh, rhessi, xrs
+        These are the keywords for the source packages currently supported by TimeSeries.
```
These changes should be undone.

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -46,11 +47,10 @@ class TimeSeriesFactory(BasicRegistrationFactory):
+        These are the keywords for the source packages currently supported by TimeSeries.
 
-    source : `str`, optional
+    source : {values}
```
We should make sure these look like: 

```
order : {'C', 'F', 'A'}
    Description of `order`.
```
from https://numpydoc.readthedocs.io/en/latest/format.html

davnish commented:
Are these changes okay? @nabobalis

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -50,7 +51,7 @@ class TimeSeriesFactory(BasicRegistrationFactory):
         detailed list of possible inputs.
 
-    source : `str`, optional
+    source : {Source}
```
`({‘eve’, ‘noaaindices’, ‘norh’, ‘rhessi’, ‘gbmsummary’, ‘lyra’, ‘xrs’, ‘esp’, ‘noaapredictindices’})`

This lacks the `|`  required as per the numpydoc style.

nabobalis commented on sunpy/timeseries/sources/__init__.py:
```diff
@@ -13,3 +13,14 @@
+__all__ = ['ESPTimeSeries', 'EVESpWxTimeSeries', 'GBMSummaryTimeSeries', 'XRSTimeSeries', 'LYRATimeSeries', 
+           'NOAAIndicesTimeSeries', 'NOAAPredictIndicesTimeSeries', 'NoRHTimeSeries', 'RHESSISummaryTimeSeries']
+
+def source_set():
+    source = set()
+
+    for i in __all__:
+       source.add(globals()[i]._source)
+
+    return source
```
Does this need to be a function?

nabobalis commented on sunpy/timeseries/sources/__init__.py:
```diff
@@ -13,3 +13,10 @@
+__all__ = ['ESPTimeSeries', 'EVESpWxTimeSeries', 'GBMSummaryTimeSeries', 'XRSTimeSeries', 'LYRATimeSeries', 
+           'NOAAIndicesTimeSeries', 'NOAAPredictIndicesTimeSeries', 'NoRHTimeSeries', 'RHESSISummaryTimeSeries']
+
+source_set = set()
+for i in __all__:
+    source_set.add(globals()[i]._source)
```
Can this not be a list comprehension?

nabobalis commented on sunpy/timeseries/sources/__init__.py:
```diff
@@ -13,3 +13,10 @@
+__all__ = ['ESPTimeSeries', 'EVESpWxTimeSeries', 'GBMSummaryTimeSeries', 'XRSTimeSeries', 'LYRATimeSeries', 
+           'NOAAIndicesTimeSeries', 'NOAAPredictIndicesTimeSeries', 'NoRHTimeSeries', 'RHESSISummaryTimeSeries']
+
+source_set = set()
```
could this be renamed to sources or source_names?

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -521,7 +522,8 @@ class NoTimeSeriesFound(ValueError):
                                default_widget_type=GenericTimeSeries,
                                additional_validation_functions=['is_datasource_for'])
+
+TimeSeriesFactory.__doc__ = TimeSeriesFactory.__doc__.format(source=source_set)
```
The formating has to add the | between each value.

davnish commented:
@nabobalis I did the changes you asked for. I am not sure about the last change, I have done the changes but is this the change your asked for?

nabobalis commented:
The precommit needs fixing.

davnish commented:
pre-commit.ci autofix

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -525,3 +526,5 @@ class NoTimeSeriesFound(ValueError):
                                default_widget_type=GenericTimeSeries,
                                additional_validation_functions=['is_datasource_for'])
+
+TimeSeriesFactory.__doc__ = TimeSeriesFactory.__doc__.format(source="{"+' | '.join({f"'{s}'" for s in source_names})+"}")
```
Woops, I am wrong. The `|` should not be there.

nabobalis commented on sunpy/timeseries/timeseries_factory.py:
```diff
@@ -50,7 +51,7 @@ class TimeSeriesFactory(BasicRegistrationFactory):
         detailed list of possible inputs.
 
-    source : `str`, optional
+    source : {source}
```
```suggestion
    source : {source}, optional 
```

nabobalis commented:
Thanks for the PR @davnish
